### PR TITLE
Bug/databricks sql incremental

### PR DESCRIPTION
**Pull request by fivetran-catfritz**
## PR Overview
**This PR will address the following Issue/Feature:**
- #50 
- Updated lookback macro what we use in Netsuite.

**This PR will result in the following new package version:**

- v0.10.0 since we are changing file formats and materializations.

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

>## 🚨 Breaking Changes 🚨
> ⚠️ Since the following changes result in the table format changing, we recommend running a `--full-refresh` after upgrading to this version to avoid possible incremental failures.

> - For Databricks All-Purpose clusters, incremental models will now be materialized using the delta table format (previously parquet).
>   - Delta tables are generally more performant than parquet and are also more widely available for Databricks users. This will also prevent compilation issues on customers' managed tables.
> 
> - For Databricks SQL Warehouses, incremental materialization will not be used due to the incompatibility of the `insert_overwrite` strategy.
> 
> ## Under the Hood
> - The `is_incremental_compatible` macro has been added to return `true` if the target warehouse supports our chosen incremental strategy.
>   - This update was applied as there have been other Databricks runtimes discovered (ie. an endpoint and external runtime) which do not support the `insert_overwrite` incremental strategy used. 
> - Added integration testing for Databricks SQL Warehouse.
> - Added consistency tests for models:
>   - `mixpanel__daily_events`
>   - `mixpanel__event`
>   - `mixpanel__monthly_events`
>   - `mixpanel__sessions`
> - Updated logic for macro `mixpanel_lookback` to align with logic used in similar macros in other packages. 

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [x] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned.
- [x] All necessary documentation and version upgrades have been applied.
    
- [x] docs were regenerated (unless this PR does not include any code or yml updates).
- [x] BuildKite integration tests are passing.
- [x] Detailed validation steps have been provided below.

### Detailed Validation
Please share any and all of your validation steps:

#### Consistency tests pass
- Screenshot 2024-07-26 at 11 22 03 AM

#### Parquet file format
- Before addressing the insert_overwrite incompatibility, our Databricks SQL instance is "managed", so I was getting the below error for the parquet file format.
  - Screenshot 2024-07-24 at 4 50 01 PM

- Updating to delta format resolves this issue.

#### Insert-overwrite incremental strategy
- Once the file format issue was resolved, I was able to reproduce the reported issue and get the below error.
  - Screenshot 2024-07-24 at 4 52 16 PM

- Updating with the `materialized='incremental' if is_incremental_compatible() else 'table'` approach resolved this issue. When running a non-full-refresh run with a target of Databricks SQL, confirm there is no error and tables are created instead of and incremental run.
  - Screenshot 2024-07-24 at 4 55 48 PM

### If you had to summarize this PR in an emoji, which would it be?
 
:dancer:

### CODE REVIEW

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-catfritz these changes look great! Just a few suggestions and one code update request which is the same as my [HubSpot comment](https://github.com/fivetran/dbt_hubspot/pull/144#discussion_r1693609828). Let me know if you have any questions. 

Once these updates are applied the re-review should be good for approval!

**Review by fivetran-catfritz — COMMENTED** [review]
@fivetran-joemarkiewicz This one is ready for re-review as well!

File: integration_tests/tests/consistency/row_comparisons/consistency_sessions.sql [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+-- this test ensures the daily_activity end model matches the prior version
+with prod as (
+    select *
+    from {{ target.schema }}_mixpanel_prod.mixpanel__sessions
+),
+
+dev as (
+    select *
+    from {{ target.schema }}_mixpanel_dev.mixpanel__sessions
+),
```
Comment (fivetran-joemarkiewicz) [inline integration_tests/tests/consistency/row_comparisons/consistency_sessions.sql:15]: The field `event_frequencies` is a result of a string agg array and the ordering of the fields can vary based on the run. Therefore, we should exclude this field from the comparison so we do not see false failures in the future.

```suggestion
{% set exclude_fields = ["event_frequencies"] %}
{% set fields = dbt_utils.star(from=ref('mixpanel__sessions'), except=exclude_fields) %}

-- this test ensures the daily_activity end model matches the prior version
with prod as (
    select {{ fields }}
    from {{ target.schema }}_mixpanel_prod.mixpanel__sessions
),

dev as (
    select {{ fields }}
    from {{ target.schema }}_mixpanel_dev.mixpanel__sessions
),
```

File: integration_tests/tests/consistency/row_comparisons/consistency_event.sql [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+-- this test ensures the daily_activity end model matches the prior version
+with prod as (
+    select *
+    from {{ target.schema }}_mixpanel_prod.mixpanel__event
+),
+
+dev as (
+    select *
+    from {{ target.schema }}_mixpanel_dev.mixpanel__event
+),
```
Comment (fivetran-joemarkiewicz) [inline integration_tests/tests/consistency/row_comparisons/consistency_event.sql:15]: The `properties` and `event_properties` fields have the potential to be interpreted as JSON/SUPER in certain warehouses (as opposed to JSON formatted string). JSON datatypes are not supported in except distinct operations. Therefore, we should remove these fields from the comparison so we do not encounter false failures in the future.

image


```suggestion
{% set exclude_fields = ["properties", 'event_properties'] %}
{% set fields = dbt_utils.star(from=ref('mixpanel__event'), except=exclude_fields) %}

-- this test ensures the daily_activity end model matches the prior version
with prod as (
    select {{ fields }}
    from {{ target.schema }}_mixpanel_prod.mixpanel__event
),

dev as (
    select {{ fields }}
    from {{ target.schema }}_mixpanel_dev.mixpanel__event
),
```

**Review by fivetran-joemarkiewicz — APPROVED** [review]
@fivetran-catfritz great work on this PR and thanks for addressing my review notes. This PR looks good for release review! I did encounter a few issues when running the validations with internal data, but they were pretty small and we have an easy workaround. Please take a look and address my remaining notes, but no need for those to be a blocker from initiating the release review process.

Great work!

File: integration_tests/tests/consistency/row_comparisons/consistency_event.sql (hunk 2 above)
Comment (fivetran-catfritz) [reply integration_tests/tests/consistency/row_comparisons/consistency_event.sql]: Thank you--updated!

File: integration_tests/tests/consistency/row_comparisons/consistency_sessions.sql (hunk 1 above)
Comment (fivetran-catfritz) [reply integration_tests/tests/consistency/row_comparisons/consistency_sessions.sql]: Thank you--added!

**Review by fivetran-catfritz — COMMENTED** [review]
Thank you! I have updated the tests, so now proceeding with release!

**Pull request merged**